elastic/apm-server
# sampling: use a data stream for sampled trace docs

## Motivation/summary

Update tail-based sampling to index into and search a data stream. The data stream will be associated with an ILM policy that takes care of rollover and deletion.

When running in Fleet-managed mode, apm-server will expect the data stream and ILM policy to exist for the data stream called `traces-sampled-<namespace>`. Servers participating in tail-based sampling are required to be configured with the same namespace.

When running in standalone mode, apm-server will attempt to create an index template and ILM policy for a data stream called `apm-sampled-traces`. This is added for minimal support while we transition things to Fleet, and is intended to be removed in a future release. The data stream is not intended to adhere to the standard indexing strategy.

## Checklist


## How to test these changes

`cd systemtest && go test -v -run TailSampling`

For manual testing:
 - see https://github.com/elastic/apm-server/issues/4584 for how to test the fix for original issue
 - run apm-server with tail sampling enabled:
   - ensure an index template and ILM policy both called "apm-sampled-traces" are created
   - ensure a data stream called "apm-sampled-traces"

It's not currently possible to configure tail sampling with Fleet, so no testing steps are described for those parts. Just check that installing the package leads to the "sampled_traces" data stream index template, ILM policy, and ingest pipeline being installed.

## Related issues

Closes #4584
